lib/internal/modules/esm/module_job.js

```diff
@@ -75,11 +77,20 @@ const isCommonJSGlobalLikeNotDefinedError = (errorMessage) =>
  * @returns {void}
  */
 const explainCommonJSGlobalLikeNotDefinedError = (e, url, hasTopLevelAwait) => {
-  if (e?.name === 'ReferenceError' &&
-      isCommonJSGlobalLikeNotDefinedError(e.message)) {
+  const undefinedGlobal = getUndefinedCJSGlobalLike(e?.message);
+  if (e?.name === 'ReferenceError' && undefinedGlobal !== null) {
 
     if (hasTopLevelAwait) {
-      e.message = `Cannot determine intended module format because both require() and top-level await are present. If the code is intended to be CommonJS, wrap await in an async function. If the code is intended to be an ES module, replace require() with import.`;
+      let advice = '';
+      if (undefinedGlobal === 'require') {
+        advice = 'replace require() with import';
+      } else if (undefinedGlobal === 'module' || undefinedGlobal === 'exports') {
+        advice = 'use export instead of module.exports/exports';
+      } else if (undefinedGlobal === '__filename' || undefinedGlobal === '__dirname') {
+        advice = 'use import.meta.url instead';
```

Comment:
[`import.meta.filename`](https://nodejs.org/api/esm.html#importmetadirname) and [`import.meta.dirname`](https://nodejs.org/api/esm.html#importmetafilename) are more direct replacements.